File: backend/src/main/java/project/backend/global/exception/GlobalExceptionHandler.java
package project.backend.global.exception;


import java.util.HashMap;
import java.util.Map;

import java.util.Optional;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;

import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import project.backend.global.exception.ex.AuthException;
import project.backend.global.exception.ex.BaseException;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

	@ExceptionHandler(BaseException.class)
	public ResponseEntity<?> handleBaseException(BaseException ex) {
		ErrorResponse response = ErrorResponse.toResponse(ex.getErrorCode());

		return ResponseEntity
			.status(ex.getStatus())
			.body(response);
	}

	@ExceptionHandler(MethodArgumentNotValidException.class)
	public ResponseEntity<?> handleMethodArgumentNotValidException(
		MethodArgumentNotValidException ex) {
		log.info("ex.getMessage() = {}", ex.getMessage());
		//필드 에러
		if (!ex.getBindingResult().getFieldErrors().isEmpty()) {
			FieldError fieldError = ex.getBindingResult().getFieldErrors().getFirst();
			return ResponseEntity
				.status(HttpStatus.BAD_REQUEST)
				.body(ErrorResponse.toResponse(fieldError));
		}

		//객체 에러(@AssertTrue 같은거)
		if (!ex.getBindingResult().getGlobalErrors().isEmpty()) {
			String message = ex.getBindingResult().getGlobalErrors().getFirst().getDefaultMessage();
			return ResponseEntity
				.status(HttpStatus.BAD_REQUEST)
				.body(ErrorResponse.builder()
					.code("VALIDATION_FAILED")
					.message(message)
					.build());
		}

		return ResponseEntity
			.status(HttpStatus.BAD_REQUEST)
			.body(ErrorResponse.builder()
				.code("UNEXPECTED_VALIDATION_FAILED")
				.message(ex.getMessage())
				.build());
	}


}

File: backend/src/main/java/project/backend/global/exception/GlobalExceptionHandler.kt
package project.backend.global.exception

import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import project.backend.global.exception.ErrorResponse.Companion.toResponse
import project.backend.global.exception.ex.BaseException

@RestControllerAdvice
class GlobalExceptionHandler {

    private val log = KotlinLogging.logger {}

    @ExceptionHandler(BaseException::class)
    fun handleBaseException(ex: BaseException): ResponseEntity<ErrorResponse> {
        return ResponseEntity
            .status(ex.status)
            .body(toResponse(ex.errorCode))
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleMethodArgumentNotValidException(
        ex: MethodArgumentNotValidException
    ): ResponseEntity<ErrorResponse> {
        log.info { "ex.getMessage() = ${ex.message}" }

        val bindingResult = ex.bindingResult

        //필드 에러
        bindingResult.fieldErrors.firstOrNull()?.let { fieldError ->
            return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(toResponse(fieldError))
        }

        //객체 에러(@AssertTrue 같은거)
        bindingResult.globalErrors.firstOrNull()?.let { globalError ->
            return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(ErrorResponse("VALIDATION_FAILED", globalError.defaultMessage ?: "검증 실패"))
        }

        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(
                ErrorResponse("UNEXPECTED_VALIDATION_FAILED", ex.message ?: "검증 실패")
            )
    }
}
